README.rst
#######
pycopic
#######


*****
About
*****
An attempt to mix Pycom's `pycoproc`_ with `pypic`_ in order to
control the PIC through the serial interface and use extended
functionalities even on the expansion board, where the PIC
is not connected to the I2C bus of the ESP32.

This is essentially a hack, YMMV.

.. note:: This is still a work-in-progress and didn't work out successfully yet.


********
Synopsis
********
::

    export MCU_PORT=/dev/cu.usbmodemPy001711

    pycopic --port $MCU_PORT --sleep 30
    pycopic --port $MCU_PORT --reset


*********
Resources
*********

.. todo:: Add some links to the Pycom user forum here.


.. _pycoproc: https://github.com/pycom/pycom-libraries/blob/master/lib/pycoproc/pycoproc.py
.. _pypic: https://github.com/pycom/pycom-micropython-sigfox/blob/Dev/esp32/tools/pypic.py


CHANGES.rst
#########
Changelog
#########


in progress
===========
- Add unmodified sources of ``pycoproc.py`` and ``pypic.py`` from Pycom.
- Properly address class variables within "pycoproc.py".
- Raise exception when Pycoproc.calibrate_rtc() fails
- Python 3 compatibility for "pypic.py"
- Add pycopic.py and setup.py
- Add .gitignore, README.rst and CHANGES.rst


pypic.py
#!/usr/bin/env python

from __future__ import print_function

import sys
import argparse
import errno
import serial
import struct
import time

__version__ = '0.9.2'

CMD_PEEK = (0x0)
CMD_POKE = (0x01)
CMD_MAGIC = (0x02)
CMD_HW_VER = (0x10)
CMD_FW_VER = (0x11)
CMD_PROD_ID = (0x12)
CMD_SETUP_SLEEP = (0x20)
CMD_GO_SLEEP = (0x21)
CMD_CALIBRATE = (0x22)
CMD_BAUD_CHANGE = (0x30)
CMD_DFU = (0x31)

ANSELA_ADDR = (0x18C)
ANSELB_ADDR = (0x18D)
ANSELC_ADDR = (0x18E)

ADCON0_ADDR = (0x9D)
ADCON1_ADDR = (0x9E)

IOCAP_ADDR = (0x391)
IOCAN_ADDR = (0x392)

_ADCON0_CHS_POSN = (0x02)
_ADCON0_ADON_MASK = (0x01)
_ADCON1_ADCS_POSN = (0x04)
_ADCON0_GO_nDONE_MASK = (0x02)

ADRESL_ADDR = (0x09B)
ADRESH_ADDR = (0x09C)

TRISC_ADDR = (0x08E)

PORTA_ADDR = (0x00C)
PORTC_ADDR = (0x00E)

WPUA_ADDR = (0x20C)

PCON_ADDR = (0x096)
STATUS_ADDR = (0x083)


# helper functions
def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)

def log(*args):
    print(' '.join(str(a) for a in args))

def error(msg):
    eprint('error:', msg)

def exit_with_error(code, msg):
    error(msg)
    sys.exit(code)

def warn(msg):
    eprint('warning:', msg)


class Pypic:

    def __init__(self, port):
        # we need bytesize to be 5 bits in order for the PIC to process the commands
        self.serial = serial.Serial(port, baudrate=115200, bytesize=serial.FIVEBITS, timeout=0.25)
        self.detected = False

        try:
            if self.read_fw_version() < 6:
                raise ValueError('PIC firmware out of date')
            else:
                self.detected = True
        except Exception:
            pass

    def _write(self, data, read=True):
        self.serial.write(data)
        if read:
            r_data = self.serial.read(2)
            if not r_data:
                raise Exception('Timeout while waiting for Rx data')
            return r_data[0]

    def _send_cmd(self, cmd):
        return self._write(bytearray([cmd]))

    def read_hw_version(self):
        return self._send_cmd(CMD_HW_VER)

    def read_fw_version(self):
        return self._send_cmd(CMD_FW_VER)

    def read_product_id(self):
        return self._send_cmd(CMD_PROD_ID)

    def peek_memory(self, addr):
        return self._write(bytearray([CMD_PEEK, addr & 0xFF, (addr >> 8) & 0xFF]))

    def poke_memory(self, addr, value):
         self._write(bytearray([CMD_POKE, addr & 0xFF, (addr >> 8) & 0xFF, value & 0xFF]), False)

    def magic_write_read(self, addr, _and=0xFF, _or=0, _xor=0):
        return self._write(bytearray([CMD_MAGIC, addr & 0xFF, (addr >> 8) & 0xFF, _and & 0xFF, _or & 0xFF, _xor & 0xFF]))

    def magic_write(self, addr, _and=0xFF, _or=0, _xor=0):
        self._write(bytearray([CMD_MAGIC, addr & 0xFF, (addr >> 8) & 0xFF, _and & 0xFF, _or & 0xFF, _xor & 0xFF]), False)

    def toggle_bits_in_memory(self, addr, bits):
        self.magic_write(addr, _xor=bits)

    def mask_bits_in_memory(self, addr, mask):
        self.magic_write(addr, _and=mask)

    def set_bits_in_memory(self, addr, bits):
        self.magic_write(addr, _or=bits)

    def reset_pycom_module(self):
        # make RC5 an output
        self.mask_bits_in_memory(TRISC_ADDR, ~(1 << 5))
        # drive RC5 low
        self.mask_bits_in_memory(PORTC_ADDR, ~(1 << 5))
        time.sleep(0.2)
        # drive RC5 high
        self.set_bits_in_memory(PORTC_ADDR, 1 << 5)
        time.sleep(0.1)
        # make RC5 an input
        self.set_bits_in_memory(TRISC_ADDR, 1 << 5)

    def enter_pycom_programming_mode(self):
        # make RC0 an output
        self.mask_bits_in_memory(TRISC_ADDR, ~(1 << 0))
        # set RC0 low
        self.mask_bits_in_memory(PORTC_ADDR, ~(1 << 0))
        # perform reset
        self.reset_pycom_module()
        # We should keep RC0 low at this point in case someone
        # presses the reset button before the firmware upgrade
        # as this is mandatory for the regular expansion board

    def exit_pycom_programming_mode(self):
        # make RC0 an output
        self.mask_bits_in_memory(TRISC_ADDR, ~(1 << 0))
        # set RC0 high
        self.set_bits_in_memory(PORTC_ADDR, 1 << 0)
        # perform reset
        self.reset_pycom_module()
        time.sleep(0.1)
        # make RC0 an input
        # This will prevent issues with the RGB LED
        self.set_bits_in_memory(TRISC_ADDR, 1 << 0)

    def isdetected(self):
        return self.detected

    def close(self):
        self.serial.close()


def main(args):
    parser = argparse.ArgumentParser(description='Sends internal commands to put the Pycom module in programming mode')
    parser.add_argument('-p', '--port', metavar='PORT', help='the serial port used to communicate with the PIC')
    parser.add_argument('--enter', action='store_true', help='enter programming mode')
    parser.add_argument('--exit', action='store_true', help='exit programming mode')
    args = parser.parse_args()

    if not args.port:
        exit_with_error(1, 'no serial port specified')

    if (args.enter and args.exit) or (not args.enter and not args.exit):
        exit_with_error(1, 'invalid action requested')

    pic = Pypic(args.port)

    if pic.isdetected():
        if args.enter:
            pic.enter_pycom_programming_mode()
        elif args.exit:
            pic.exit_pycom_programming_mode()

    pic.close()


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))


pycoproc.py
#!/usr/bin/env python
#
# Copyright (c) 2020, Pycom Limited.
#
# This software is licensed under the GNU GPL version 3 or any
# later version, with permitted additional terms. For more information
# see the Pycom Licence v1.0 document supplied with this file, or
# available at https://www.pycom.io/opensource/licensing
#

# See https://docs.pycom.io for more information regarding library specifics

from machine import Pin
from machine import I2C
import time
import pycom

__version__ = '0.0.2'

""" PIC MCU wakeup reason types """
WAKE_REASON_ACCELEROMETER = 1
WAKE_REASON_PUSH_BUTTON = 2
WAKE_REASON_TIMER = 4
WAKE_REASON_INT_PIN = 8

class Pycoproc:
    """ class for handling the interaction with PIC MCU """

    I2C_SLAVE_ADDR = const(8)

    PYSENSE = const(1)
    PYTRACK = const(2)
    PYSCAN = const(3)

    BOARD_TYPE_SET = (PYSENSE, PYTRACK, PYSCAN)

    CMD_PEEK = const(0x0)
    CMD_POKE = const(0x01)
    CMD_MAGIC = const(0x02)
    CMD_HW_VER = const(0x10)
    CMD_FW_VER = const(0x11)
    CMD_PROD_ID = const(0x12)
    CMD_SETUP_SLEEP = const(0x20)
    CMD_GO_SLEEP = const(0x21)
    CMD_CALIBRATE = const(0x22)
    CMD_BAUD_CHANGE = const(0x30)
    CMD_DFU = const(0x31)

    REG_CMD = const(0)
    REG_ADDRL = const(1)
    REG_ADDRH = const(2)
    REG_AND = const(3)
    REG_OR = const(4)
    REG_XOR = const(5)

    ANSELA_ADDR = const(0x18C)
    ANSELB_ADDR = const(0x18D)
    ANSELC_ADDR = const(0x18E)

    ADCON0_ADDR = const(0x9D)
    ADCON1_ADDR = const(0x9E)

    IOCAP_ADDR = const(0x391)
    IOCAN_ADDR = const(0x392)

    INTCON_ADDR = const(0x0B)
    OPTION_REG_ADDR = const(0x95)

    _ADCON0_CHS_POSN = const(0x02)
    _ADCON0_ADON_MASK = const(0x01)
    _ADCON1_ADCS_POSN = const(0x04)
    _ADCON0_GO_nDONE_MASK = const(0x02)

    ADRESL_ADDR = const(0x09B)
    ADRESH_ADDR = const(0x09C)

    TRISC_ADDR = const(0x08E)

    PORTA_ADDR = const(0x00C)
    PORTC_ADDR = const(0x00E)

    WPUA_ADDR = const(0x20C)

    WAKE_REASON_ADDR = const(0x064C)
    MEMORY_BANK_ADDR = const(0x0620)

    PCON_ADDR = const(0x096)
    STATUS_ADDR = const(0x083)

    EXP_RTC_PERIOD = const(7000)

    def __init__(self, board_type, i2c=None, sda='P22', scl='P21'):
        if i2c is not None:
            self.i2c = i2c
        else:
            self.i2c = I2C(0, mode=I2C.MASTER, pins=(sda, scl))

        if board_type not in self.BOARD_TYPE_SET:
            raise Exception('Board type not in the set {}'.format(self.BOARD_TYPE_SET))

        self.sda = sda
        self.scl = scl
        self.board_type = board_type
        self.clk_cal_factor = 1
        self.reg = bytearray(6)
        self.wake_int = False
        self.wake_int_pin = False
        self.wake_int_pin_rising_edge = True

        # Make sure we are inserted into the
        # correct board and can talk to the PIC
        try:
            self.read_fw_version()
        except Exception as e:
            raise Exception('Board not detected: {}'.format(e))

        # init the ADC for the battery measurements
        self.poke_memory(self.ANSELC_ADDR, 1 << 2)
        self.poke_memory(self.ADCON0_ADDR, (0x06 << self._ADCON0_CHS_POSN) | self._ADCON0_ADON_MASK)
        self.poke_memory(self.ADCON1_ADDR, (0x06 << self._ADCON1_ADCS_POSN))
        # enable the pull-up on RA3
        self.poke_memory(self.WPUA_ADDR, (1 << 3))
        # make RC5 an input
        self.set_bits_in_memory(self.TRISC_ADDR, 1 << 5)
        # set RC6 and RC7 as outputs and enable power to the sensors and the GPS
        self.mask_bits_in_memory(self.TRISC_ADDR, ~(1 << 6))
        self.mask_bits_in_memory(self.TRISC_ADDR, ~(1 << 7))

        if self.read_fw_version() < 6:
            raise ValueError('Firmware out of date')


    def _write(self, data, wait=True):
        self.i2c.writeto(self.I2C_SLAVE_ADDR, data)
        if wait:
            self._wait()

    def _read(self, size):
        return self.i2c.readfrom(self.I2C_SLAVE_ADDR, size + 1)[1:(size + 1)]

    def _wait(self):
        count = 0
        time.sleep_us(10)
        while self.i2c.readfrom(self.I2C_SLAVE_ADDR, 1)[0] != 0xFF:
            time.sleep_us(100)
            count += 1
            if (count > 500):  # timeout after 50ms
                raise Exception('Board timeout')

    def _send_cmd(self, cmd):
        self._write(bytes([cmd]))

    def read_hw_version(self):
        self._send_cmd(self.CMD_HW_VER)
        d = self._read(2)
        return (d[1] << 8) + d[0]

    def read_fw_version(self):
        self._send_cmd(self.CMD_FW_VER)
        d = self._read(2)
        return (d[1] << 8) + d[0]

    def read_product_id(self):
        self._send_cmd(self.CMD_PROD_ID)
        d = self._read(2)
        return (d[1] << 8) + d[0]

    def peek_memory(self, addr):
        self._write(bytes([self.CMD_PEEK, addr & 0xFF, (addr >> 8) & 0xFF]))
        return self._read(1)[0]

    def poke_memory(self, addr, value):
        self._write(bytes([self.CMD_POKE, addr & 0xFF, (addr >> 8) & 0xFF, value & 0xFF]))

    def magic_write_read(self, addr, _and=0xFF, _or=0, _xor=0):
        self._write(bytes([self.CMD_MAGIC, addr & 0xFF, (addr >> 8) & 0xFF, _and & 0xFF, _or & 0xFF, _xor & 0xFF]))
        return self._read(1)[0]

    def toggle_bits_in_memory(self, addr, bits):
        self.magic_write_read(addr, _xor=bits)

    def mask_bits_in_memory(self, addr, mask):
        self.magic_write_read(addr, _and=mask)

    def set_bits_in_memory(self, addr, bits):
        self.magic_write_read(addr, _or=bits)

    def get_wake_reason(self):
        """ returns the wakeup reason, a value out of constants WAKE_REASON_* """
        return self.peek_memory(self.WAKE_REASON_ADDR)

    def get_sleep_remaining(self):
        """ returns the remaining time from sleep, as an interrupt (wakeup source) might have triggered """
        c3 = self.peek_memory(self.WAKE_REASON_ADDR + 3)
        c2 = self.peek_memory(self.WAKE_REASON_ADDR + 2)
        c1 = self.peek_memory(self.WAKE_REASON_ADDR + 1)
        time_device_s = (c3 << 16) + (c2 << 8) + c1
        # this time is from PIC internal oscilator, so it needs to be adjusted with the calibration value
        try:
            self.calibrate_rtc()
        except Exception:
            pass
        time_s = int((time_device_s / self.clk_cal_factor) + 0.5) # 0.5 used for round
        return time_s

    def setup_sleep(self, time_s):
        try:
            self.calibrate_rtc()
        except Exception:
            raise
        time_s = int((time_s * self.clk_cal_factor) + 0.5)  # round to the nearest integer
        if time_s >= 2**(8*3):
            time_s = 2**(8*3)-1
        self._write(bytes([self.CMD_SETUP_SLEEP, time_s & 0xFF, (time_s >> 8) & 0xFF, (time_s >> 16) & 0xFF]))

    def go_to_sleep(self, gps=True):
        # if we have a Pytrack then enable or disable back-up power to the GPS receiver
        if self.board_type == self.PYTRACK and gps:
            # disable GPS only if Pytrack
            self.set_bits_in_memory(self.PORTC_ADDR, 1 << 7)
        else:
            # Pysense or Pyscan or no GPS
            self.mask_bits_in_memory(self.PORTC_ADDR, ~(1 << 7))

        # disable the ADC
        self.poke_memory(self.ADCON0_ADDR, 0)

        if self.wake_int:
            # Don't touch RA3, RA5 or RC1 so that interrupt wake-up works
            self.poke_memory(self.ANSELA_ADDR, ~((1 << 3) | (1 << 5)))
            self.poke_memory(self.ANSELC_ADDR, ~((1 << 6) | (1 << 7) | (1 << 1)))
        else:
            # disable power to the accelerometer, and don't touch RA3 so that button wake-up works
            self.poke_memory(self.ANSELA_ADDR, ~(1 << 3))
            self.poke_memory(self.ANSELC_ADDR, ~(1 << 7))

        self.poke_memory(self.ANSELB_ADDR, 0xFF)

        # check if INT pin (PIC RC1), should be used for wakeup
        if self.wake_int_pin:
            if self.wake_int_pin_rising_edge:
                self.set_bits_in_memory(self.OPTION_REG_ADDR, 1 << 6) # rising edge of INT pin
            else:
                self.mask_bits_in_memory(self.OPTION_REG_ADDR, ~(1 << 6)) # falling edge of INT pin
            self.mask_bits_in_memory(self.ANSELC_ADDR, ~(1 << 1)) # disable analog function for RC1 pin
            self.set_bits_in_memory(self.TRISC_ADDR, 1 << 1) # make RC1 input pin
            self.mask_bits_in_memory(self.INTCON_ADDR, ~(1 << 1)) # clear INTF
            self.set_bits_in_memory(self.INTCON_ADDR, 1 << 4) # enable interrupt; set INTE)

        self._write(bytes([self.CMD_GO_SLEEP]), wait=False)
        # kill the run pin
        Pin('P3', mode=Pin.OUT, value=0)

    def calibrate_rtc(self):
        # the 1.024 factor is because the PIC LF operates at 31 KHz
        # WDT has a frequency divider to generate 1 ms
        # and then there is a binary prescaler, e.g., 1, 2, 4 ... 512, 1024 ms
        # hence the need for the constant
        self._write(bytes([self.CMD_CALIBRATE]), wait=False)
        self.i2c.deinit()
        Pin('P21', mode=Pin.IN)
        pulses = pycom.pulses_get('P21', 100)
        self.i2c.init(mode=I2C.MASTER, pins=(self.sda, self.scl))
        idx = 0
        for i in range(len(pulses)):
            if pulses[i][1] > self.EXP_RTC_PERIOD:
                idx = i
                break
        try:
            period = pulses[idx][1] - pulses[(idx - 1)][1]
        except:
            period = 0
        if period > 0:
            self.clk_cal_factor = (self.EXP_RTC_PERIOD / period) * (1000 / 1024)
        if self.clk_cal_factor > 1.25 or self.clk_cal_factor < 0.75:
            self.clk_cal_factor = 1

    def button_pressed(self):
        button = self.peek_memory(self.PORTA_ADDR) & (1 << 3)
        return not button

    def read_battery_voltage(self):
        self.set_bits_in_memory(self.ADCON0_ADDR, self._ADCON0_GO_nDONE_MASK)
        time.sleep_us(50)
        while self.peek_memory(self.ADCON0_ADDR) & self._ADCON0_GO_nDONE_MASK:
            time.sleep_us(100)
        adc_val = (self.peek_memory(self.ADRESH_ADDR) << 2) + (self.peek_memory(self.ADRESL_ADDR) >> 6)
        return (((adc_val * 3.3 * 280) / 1023) / 180) + 0.01    # add 10mV to compensate for the drop in the FET

    def setup_int_wake_up(self, rising, falling):
        """ rising is for activity detection, falling for inactivity """
        wake_int = False
        if rising:
            self.set_bits_in_memory(self.IOCAP_ADDR, 1 << 5)
            wake_int = True
        else:
            self.mask_bits_in_memory(self.IOCAP_ADDR, ~(1 << 5))

        if falling:
            self.set_bits_in_memory(self.IOCAN_ADDR, 1 << 5)
            wake_int = True
        else:
            self.mask_bits_in_memory(self.IOCAN_ADDR, ~(1 << 5))
        self.wake_int = wake_int

    def setup_int_pin_wake_up(self, rising_edge = True):
        """ allows wakeup to be made by the INT pin (PIC -RC1) """
        self.wake_int_pin = True
        self.wake_int_pin_rising_edge = rising_edge


pycopic.py
import sys
import argparse
import time

from mock import MagicMock
from pypic import Pypic


def patch_python():

    import builtins
    builtins.const = int

    sys.modules['pycom'] = MagicMock()
    sys.modules['machine'] = MagicMock()


def patch_pycoproc():

    global PycoprocForSerial

    from pycoproc import Pycoproc
    class PycoprocForSerial(Pycoproc):

        def __init__(self, pypic, *args, **kwargs):
            self.pypic = pypic
            self.board_type = None
            self.clk_cal_factor = 1
            self.reg = bytearray(6)
            self.wake_int = True
            self.wake_int_pin = False
            self.wake_int_pin_rising_edge = True

        def _write(self, data, wait=True):
            self.pypic._write(data, read=False)
            if wait:
                time.sleep(100 / 1000.0 / 1000.0)

        def _read(self, size):
            return [None]
            #return self.pypic._read(size)
            r_data = self.pypic.serial.read(2)
            #if not r_data:
            #    raise Exception('Timeout while waiting for Rx data')
            if r_data:
                return [r_data[0]]
            else:
                return [None]

        def calibrate_rtc___(self):
            # the 1.024 factor is because the PIC LF operates at 31 KHz
            # WDT has a frequency divider to generate 1 ms
            # and then there is a binary prescaler, e.g., 1, 2, 4 ... 512, 1024 ms
            # hence the need for the constant
            self._write(bytes([self.CMD_CALIBRATE]), wait=False)

        def setup_sleep(self, time_s):
            time_s = int((time_s * self.clk_cal_factor) + 0.5)  # round to the nearest integer
            if time_s >= 2 ** (8 * 3):
                time_s = 2 ** (8 * 3) - 1
            self._write(bytes([self.CMD_SETUP_SLEEP, time_s & 0xFF, (time_s >> 8) & 0xFF, (time_s >> 16) & 0xFF]))

        def go_to_sleep__(self, gps=True):
            self.poke_memory(self.ANSELB_ADDR, 0xFF)
            self._write(bytes([self.CMD_GO_SLEEP]), wait=False)


def main():

    # Monkeypatch Pycom modules.
    patch_python()
    patch_pycoproc()

    # Parse command line arguments.
    parser = argparse.ArgumentParser(description='Sends commands to the PIC on Pycom modules')
    parser.add_argument('-p', '--port', metavar='PORT', required=True, help='The serial port used to communicate with the PIC')
    parser.add_argument('-s', '--sleep', type=int, metavar='SLEEP', help='Send device to sleep mode for N seconds')
    parser.add_argument('-r', '--reset', action='store_true', help='Reset device')
    args = parser.parse_args()

    pypic = Pypic(args.port)
    pycoproc = PycoprocForSerial(pypic)

    if pypic.isdetected():
        if args.sleep:
            pycoproc.setup_sleep(args.sleep)
            pycoproc.go_to_sleep(gps=False)
        elif args.reset:
            pypic.reset_pycom_module()
    else:
        raise KeyError('No PIC detected')

    pypic.close()


if __name__ == "__main__":
    main()
